### ISSUE TITLE
Checking $? in pipeline

### ISSUE THREAD

**Issue by lark047** [koalaman/shellcheck#1993]
- [x] shellcheck.net (i.e. the latest commit) currently gives no useful warnings about this
- [x] I searched through https://github.com/koalaman/shellcheck/issues and didn't find anything related

#### Here's a snippet or screenshot that shows the problem:

```sh

#!/bin/bash

some_important_function() { echo hello; }

some_important_function | tee $(mktemp)
[[ $? -eq 0 ]] || printf 'FAILED!\n'
```

#### Here's what shellcheck currently says:

```
[[ $? -eq 0 ]] || printf 'FAILED!\n'
   ^-- SC2181: Check exit code directly with e.g. 'if mycmd;', not indirectly with $?.
```

#### Here's what I wanted or expected to see:

In this example I don't care about the status of `tee`, but running `cmd .. | tee ...` and checking `$?` will give me the status of the latter. It would be useful to show a warning about using `$?` to check the status of potentially the wrong command.

I could use

- `set -o pipefail` or
- `${PIPESTATUS[0]}`

to capture the result of the command I'm most interested in.

**Comment by koalaman** [comment]
This would be very useful, but how would ShellCheck know which command you care about? Is this specific to `tee`?

**Comment by lark047** [comment]
I agree that ShellCheck can't know the intention here (`tee` was just an example; this applies to any pipeline). I would rather have a `# shellcheck disable=SCXXXX` to show intentionality to future maintainers rather than have it be ambiguous. Consider the same example:
```
[...]
some_important_function
[[ $? -eq 0 ]] || printf 'FAILED!\n'
```
This will test the correct (only) status. Then a month later when the developer decides to add logging
```
[...]
some_important_function | tee $(mktemp)
[[ $? -eq 0 ]] || printf 'FAILED!\n'
```
With the `tee` example it's virtually certain that the original developer made the change without realizing the implications of the test on the next line. This isn't perfect as it violates [SC2181](https://github.com/koalaman/shellcheck/wiki/SC2181), but fixing the code to remove that warning suffers the same problem.
```
if ! some_important_function | tee $(mktemp); then
    printf 'FAILED!\n'
fi
```
Unless the `pipefail` option is set, I won't see FAILED! unless tee fails (and the return code of `some_important_function` is lost).

I think adding a warning here to either use `$PIPESTATUS[@]` or `set -o pipefail` would be useful. Indeed, I just fixed a production bug that resulted from this kind of error -- and there are many more instances across my codebase. 👎
